### PR TITLE
Profile allocations and track allocation mode in libPAS

### PR DESCRIPTION
**Pull request by ddegazio**
#### ce0c25336f5320048b444ac8300efec568b437d5
<pre>
Profile allocations and track allocation mode in libPAS
<a href="https://bugs.webkit.org/show_bug.cgi?id=270592">https://bugs.webkit.org/show_bug.cgi?id=270592</a>
<a href="https://rdar.apple.com/122419407">rdar://122419407</a>

Reviewed by Michael Saboff.

Adds more invocations of PAS_PROFILE to different allocation pathways
in libPAS, and passes them a new pas_allocation_mode discriminator
that threads through whether or not the allocation is for a compactible
type from the bmalloc/WTF layer.

* Source/WTF/wtf/FastMalloc.cpp:
* Source/WTF/wtf/Gigacage.cpp:
* Source/WTF/wtf/JSValueMalloc.cpp:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/CompactAllocationMode.h:
* Source/bmalloc/bmalloc/IsoHeap.cpp:
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoMallocFallback.cpp:
* Source/bmalloc/bmalloc/IsoMallocFallback.h:
* Source/bmalloc/bmalloc/bmalloc.cpp:
* Source/bmalloc/bmalloc/bmalloc.h:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap.c:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap.h:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap_inlines.h:
* Source/bmalloc/libpas/src/libpas/jit_heap.c:
* Source/bmalloc/libpas/src/libpas/pas_allocation_mode.h:
* Source/bmalloc/libpas/src/libpas/pas_bitfit_allocator_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_bitfit_page_config.h:
* Source/bmalloc/libpas/src/libpas/pas_bitfit_page_config_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_bitfit_page_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_compact_bootstrap_free_heap.c:
* Source/bmalloc/libpas/src/libpas/pas_compact_heap_reservation.c:
* Source/bmalloc/libpas/src/libpas/pas_config.h:
* Source/bmalloc/libpas/src/libpas/pas_config.h:
* Source/bmalloc/libpas/src/libpas/pas_create_basic_heap_page_caches_with_reserved_memory.c:
* Source/bmalloc/libpas/src/libpas/pas_debug_heap.h:
* Source/bmalloc/libpas/src/libpas/pas_heap_config.h:
* Source/bmalloc/libpas/src/libpas/pas_heap_config_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_immortal_heap.c:
* Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
* Source/bmalloc/libpas/src/libpas/pas_large_heap.h:
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_page_header_table.c:
* Source/bmalloc/libpas/src/libpas/pas_page_header_table.h:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_segregated_heap.c:
* Source/bmalloc/libpas/src/libpas/pas_segregated_page_config.h:
* Source/bmalloc/libpas/src/libpas/pas_segregated_page_config_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_segregated_size_directory.c:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate.h:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_array.h:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_common.h:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_intrinsic.h:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_primitive.h:
* Source/bmalloc/libpas/src/libpas/pas_try_reallocate.h:
* Source/bmalloc/libpas/src/libpas/pas_utility_heap.c:
* Source/bmalloc/libpas/src/libpas/pas_utils.h:

Canonical link: <a href="https://commits.webkit.org/275812@main">https://commits.webkit.org/275812@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/918fb2f801fec4c3a31fdb90bb494b02112a2072

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42887 "467 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21908 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45288 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45505 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39012 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19288 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35496 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43460 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18965 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36931 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16468 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16573 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37989 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/943 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/36343 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39095 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38324 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47052 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/42514 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17714 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14608 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/42262 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19339 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19503 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/49524 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5810 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18969 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10005 "Passed tests") |

### CODE REVIEW

File: Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h [hunk 1]
```diff
@@ -1487,12 +1493,15 @@ pas_local_allocator_try_allocate_with_free_bits(
...
     }
+
+    PAS_PROFILE(LOCAL_FREEBITS_ALLOCATION, result, allocator->object_size, mode);
     
     return pas_allocation_result_create_success(result);
 }
 
 static PAS_ALWAYS_INLINE pas_allocation_result
 pas_local_allocator_try_allocate_inline_cases(pas_local_allocator* allocator,
+                                              pas_allocation_mode mode,
```
Comment (msaboff) [inline Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:1504]: nit: Maybe change name from "mode" to "allocation_mode".

**Review by msaboff — APPROVED** [review]
r=me

nit: Maybe change variable names from "mode" to "allocation_mode" in all places where the short name is used.  Other places already use the more descriptive name.

**Comment by ddegazio** [comment]
Latest push resolves Michael's comment (renames `mode` -> `allocation_mode`), fixes a typo in a previous build fix (`CompactAllocationModeAllocationMode` -> `CompactAllocationMode`), and fixes a few style errors that crept in (header ordering in `bmalloc.h` and `IsoMallocFallback.h`). Style is expected to fail since this patch mostly involves libPAS, but I'm keeping an eye on the results to make sure we don't have any style errors outside of libPAS or direct libPAS API users.

**Comment by ddegazio** [comment]
Latest push should fix watchOS build, and also fix the build if TZone is enabled.

**Comment by ddegazio** [comment]
A few EWS tests are failing, but this seems to be due to infra issues? And they all passed on a previous iteration. I'm going to proceed with landing.

**Pull request merged**